File: Sources/App/Classes/Views/Input Text Field/TVCValidatedTextField.m
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\___/_/\_\\__|\__,_|\__,_|_|
 *
 * Copyright (c) 2010 - 2018 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  * Neither the name of Textual, "Codeux Software, LLC", nor the
 *    names of its contributors may be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *********************************************************************** */

#import "TLOLocalization.h"
#import "TVCErrorMessagePopoverControllerPrivate.h"
#import "TVCValidatedTextField.h"

NS_ASSUME_NONNULL_BEGIN

@interface TVCValidatedTextField ()
/* Maintain cached value so that the drawing does not call
 the validation block every time that it is called. */
@property(nonatomic, assign) BOOL cachedValidValue;
@property(nonatomic, assign) BOOL validationPerformed;
@property(nonatomic, copy, nullable, readwrite) NSString *lastValidationErrorDescription;
@end

@implementation TVCValidatedTextField

#pragma mark -
#pragma mark Public API (Normal Text Field)

- (void)awakeFromNib
{
	[super awakeFromNib];

	self.cachedValidValue = NO;
}

- (void)setCachedValidValue:(BOOL)cachedValidValue
{
	self->_cachedValidValue = cachedValidValue;

	[self _updateBackgroundForValidity];
}

/* An invalid value is signalled by a faint red tint behind the text
 in addition to the error popover. */
- (void)_updateBackgroundForValidity
{
	if (self.cachedValidValue || self.validationPerformed == NO) {
		self.drawsBackground = NO;
		self.backgroundColor = [NSColor textBackgroundColor];
	} else {
		self.drawsBackground = YES;
		self.backgroundColor = [[NSColor systemRedColor] colorWithAlphaComponent:0.08];
	}
}

- (void)dealloc
{
	[self closeValidationErrorPopover];
}

- (NSString *)value
{
	NSString *stringValue = self.stringValue;

	if (self.stringValueUsesOnlyFirstToken) {
		stringValue = stringValue.trimAndGetFirstToken;
	} else if (self.stringValueIsTrimmed) {
		stringValue = stringValue.trim;
	}

	if (stringValue.length == 0) {
		if (self.defaultValue && self.stringValueIsInvalidOnEmpty == NO) {
			return self.defaultValue;
		}
	}

	return stringValue;
}

- (NSString *)lowercaseValue
{
	return self.value.lowercaseString;
}

- (NSString *)uppercaseValue
{
	return self.value.uppercaseString;
}

- (NSInteger)integerValue
{
	return self.value.integerValue;
}

- (void)setIntegerValue:(NSInteger)integerValue
{
	self.stringValue = [NSString stringWithInteger:integerValue];
}

- (BOOL)valueIsEmpty
{
	return (self.stringValue.length == 0);
}

- (BOOL)valueIsValid
{
	return self.cachedValidValue;
}

#pragma mark -
#pragma mark Interval Validation

- (void)textDidChange:(NSNotification *)notification
{
	[self _valueChangedAction];
}

- (void)setStringValue:(NSString *)stringValue
{
	super.stringValue = stringValue;

	[self _valueChangedAction];
}

- (void)_valueChangedAction
{
	[self performValidation];

	[self _valueChangedActionPostflight];
}

- (void)_valueChangedActionPostflight
{
	[self closeValidationErrorPopover];

	[self informCallbackTextDidChange];
}

- (void)informCallbackTextDidChange
{
	if (self.textDidChangeCallback == nil) {
		return;
	}

	if ([self.textDidChangeCallback respondsToSelector:@selector(validatedTextFieldTextDidChange:)]) {
		[self.textDidChangeCallback performSelector:@selector(validatedTextFieldTextDidChange:) withObject:self];
	}
}

- (void)performValidation
{
	NSString *stringToValidate = self.stringValue;

	NSString *errorDescription = nil;

	if (stringToValidate.length > 0) {
		if (self.validationBlock) {
			errorDescription = self.validationBlock(stringToValidate);
		}
	} else {
		if (self.performValidationWhenEmpty) {
			errorDescription = self.validationBlock(stringToValidate);
		} else if (self.stringValueIsInvalidOnEmpty) {
			errorDescription = TXTLS(@"BasicLanguage[fo8-1h]");
		}
	}

	self.validationPerformed = YES;

	self.lastValidationErrorDescription = errorDescription;

	self.cachedValidValue = (errorDescription == nil);
}

- (BOOL)showValidationErrorPopover
{
	if (self.validationPerformed == NO) {
		[self performValidation];
	}

	NSString *errorDescription = self.lastValidationErrorDescription;

	if (errorDescription == nil) {
		return NO;
	}

	if (self.window == nil) {
		return NO;
	}

	[[TVCErrorMessagePopoverController sharedController] showMessage:errorDescription forView:self];

	return YES;
}

- (void)closeValidationErrorPopover
{
	[[TVCErrorMessagePopoverController sharedController] closeMessageForView:self];
}

- (void)viewWillMoveToWindow:(nullable NSWindow *)newWindow
{
	/* While outside logic is responsible for displaying the
	 validation error message, it would not hurt to help it a
	 little bit by dismissing the message when the view is no
	 longer on a window. */
	if (newWindow == nil) {
		[self closeValidationErrorPopover];
	}
}

@end

#pragma mark -
#pragma mark Text Field Cell

@implementation TVCValidatedTextFieldCell
/* The erroneous value tint is applied by the control itself through
 -backgroundColor. The cell subclass remains so nibs that name it keep
 loading without a fallback warning. */
@end

NS_ASSUME_NONNULL_END

File: Sources/App/Modules/AppKitSupport/TVCValidatedTextField.swift
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \\ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\\___/_/\\_\\__|\\__,_|\\__,_|_|
 *
 * Copyright (c) 2010 - 2026 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 *********************************************************************** */

import AppKit
import CocoaExtensions

@objc(TVCValidatedTextField)
public final class ValidatedTextField: NSTextField {
	@objc public var validationBlock: ((String) -> String?)?
	@objc public var stringValueUsesOnlyFirstToken = false
	@objc public var stringValueIsTrimmed = false
	@objc public var stringValueIsInvalidOnEmpty = false
	@objc public var performValidationWhenEmpty = false
	@objc public weak var textDidChangeCallback: AnyObject?
	@objc public var defaultValue: String?

	private var cachedValidValue = false
	private var validationPerformed = false
	@objc public private(set) var lastValidationErrorDescription: String?

	@objc public var value: String {
		var processedValue = super.stringValue

		if stringValueUsesOnlyFirstToken {
			processedValue = processedValue.firstToken
		} else if stringValueIsTrimmed {
			processedValue = processedValue.trimmingCharacters(in: .whitespacesAndNewlines)
		}

		if processedValue.isEmpty {
			if let defaultValue, stringValueIsInvalidOnEmpty == false {
				return defaultValue
			}
		}

		return processedValue
	}

	@objc public var lowercaseValue: String {
		value.lowercased()
	}

	@objc public var uppercaseValue: String {
		value.uppercased()
	}

	@objc public var valueIsEmpty: Bool {
		stringValue.isEmpty
	}

	@objc public var valueIsValid: Bool {
		cachedValidValue
	}

	/* ISOLATION-EXCEPTION: `NSObject.awakeFromNib()` is declared nonisolated, so the
	 override cannot be main-actor isolated. AppKit decodes nibs on the main thread
	 only, which is what makes the assumption safe. */
	override public nonisolated func awakeFromNib() {
		super.awakeFromNib()
		MainActor.assumeIsolated {
			cachedValidValue = false
		}
	}

	/** Isolated so the popover teardown runs on the main actor whichever thread
	 drops the last reference. */
	isolated deinit {
		closeValidationErrorPopover()
	}

	override public var stringValue: String {
		get {
			super.stringValue
		}
		set {
			super.stringValue = newValue
			valueChangedAction()
		}
	}

	override public func textDidChange(_ notification: Notification) {
		super.textDidChange(notification)
		valueChangedAction()
	}

	override public func viewWillMove(toWindow newWindow: NSWindow?) {
		super.viewWillMove(toWindow: newWindow)

		if newWindow == nil {
			closeValidationErrorPopover()
		}
	}

	override public var integerValue: Int {
		get {
			Int(value) ?? 0
		}
		set {
			stringValue = "\(newValue)"
		}
	}

	@objc public func performValidation() {
		let stringToValidate = stringValue
		var errorDescription: String?

		if stringToValidate.isEmpty == false {
			errorDescription = validationBlock?(stringToValidate)
		} else if performValidationWhenEmpty {
			errorDescription = validationBlock?(stringToValidate)
		} else if stringValueIsInvalidOnEmpty {
			errorDescription = ApplicationStrings.requiredField
		}

		validationPerformed = true
		lastValidationErrorDescription = errorDescription
		cachedValidValue = errorDescription == nil
		updateBackgroundForValidity()
	}

	@objc @discardableResult
	public func showValidationErrorPopover() -> Bool {
		if validationPerformed == false {
			performValidation()
		}

		guard let errorDescription = lastValidationErrorDescription else {
			return false
		}

		guard window != nil else {
			return false
		}

		ErrorMessagePopoverController.sharedController().showMessage(errorDescription, for: self)
		return true
	}

	@objc public func closeValidationErrorPopover() {
		ErrorMessagePopoverController.sharedController().closeMessage(for: self)
	}

	private func valueChangedAction() {
		performValidation()
		closeValidationErrorPopover()
		informCallbackTextDidChange()
	}

	private func updateBackgroundForValidity() {
		if cachedValidValue || validationPerformed == false {
			drawsBackground = false
			backgroundColor = .textBackgroundColor
			setAccessibilityHelp(nil)
		} else {
			drawsBackground = true
			backgroundColor = NSColor.systemRed.withAlphaComponent(0.08)
			setAccessibilityHelp(lastValidationErrorDescription)
		}
	}

	private func informCallbackTextDidChange() {
		guard let textDidChangeCallback else {
			return
		}

		let selector = NSSelectorFromString("validatedTextFieldTextDidChange:")

		if textDidChangeCallback.responds(to: selector) {
			_ = textDidChangeCallback.perform(selector, with: self)
		}
	}
}

@objc(TVCValidatedTextFieldCell)
public final class ValidatedTextFieldCell: NSTextFieldCell {}
